File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Views/ProgressView.kt
package com.abracode.actionui.Views

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text as M3Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUIModel
import com.abracode.actionui.Common.ActionUIValueType
import com.abracode.actionui.Common.ActionUIViewConstruction
import com.abracode.actionui.Common.ActionUILogger
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Common.LoggerLevel
import com.abracode.actionui.Common.LocalWindowModel
import com.abracode.actionui.Helpers.LocalActionUITint
import com.abracode.actionui.Helpers.numberProperty
import com.abracode.actionui.Helpers.stringProperty
import com.abracode.actionui.Helpers.LocalActionUIEnabled

/**
 * Renders a determinate progress bar or an indeterminate spinner.
 *
 * Mirror of the Apple `ProgressView` element (`ActionUI/Views/ProgressView.swift`),
 * which wraps `SwiftUI.ProgressView`. The element is **determinate** when a valid
 * `value` is supplied (and `value <= total`), otherwise **indeterminate**:
 *
 *   * determinate   -> Material3 [LinearProgressIndicator] (a bar), matching
 *     SwiftUI's default `ProgressView(value:total:)` rendering.
 *   * indeterminate -> Material3 [CircularProgressIndicator] (a spinner),
 *     matching SwiftUI's `ProgressView()` / iOS `.progressViewStyle(.circular)`.
 *
 * `progressViewStyle` overrides that pairing in either direction, mirroring the
 * Apple property of the same name: `"linear"` gives a bar even when the progress
 * is indeterminate - Material's indeterminate [LinearProgressIndicator], the
 * counterpart of SwiftUI's `.progressViewStyle(.linear)` - and `"circular"`
 * gives a ring even when it is determinate.
 *
 * **Supported properties.**
 *   * `value` - current progress, `0.0...total`. Omit (or supply an invalid /
 *     out-of-range value) for an indeterminate spinner.
 *   * `total` - maximum; defaults to `1.0` when `value` is present. Must be `> 0`.
 *   * `title` - optional label rendered above the indicator (in a [Column]).
 *   * `progressViewStyle` - `"automatic"` (the pairing above), `"linear"` or
 *     `"circular"`. An unrecognized value warns and falls back to automatic.
 *   * `actionID` - dispatched through [ActionUIModel] on tap, like `Button`.
 *   * plus the universal modifiers resolved by `applyCommonProperties`, applied
 *     to the enclosing [Column] via [modifier].
 *
 * **Value bridge.** Like Apple, `ProgressView` declares [ActionUIValueType.DOUBLE]
 * (Apple's `Double?`): a host can drive the progress at runtime via
 * `ActionUIModel.setElementValue(id, <Double>)`. As on Apple, the
 * `states["progress"]` override (set with `setElementState`) wins over the value
 * bridge, which in turn wins over the static `value` property; setting the value
 * `null` (no override and no property) reverts to the indeterminate spinner.
 * `initialValue` seeds from a pre-set runtime value or the `value` property,
 * matching Apple's `ProgressView.swift`. Static usage is unchanged.
 *
 * Sample JSON:
 * ```
 * { "type": "ProgressView", "properties": { "value": 0.5, "total": 1.0, "title": "Loading" } }
 * { "type": "ProgressView" }   // indeterminate spinner
 * { "type": "ProgressView", "properties": { "progressViewStyle": "linear" } }  // indeterminate BAR
 * ```
 */
object ProgressView : ActionUIViewConstruction {
    /** Apple's `states["progress"]` key: a runtime override of the displayed progress. */
    private const val PROGRESS_STATE_KEY = "progress"

    override val valueType = ActionUIValueType.DOUBLE

    // Non-null only when a runtime value was set; null means "use the static
    // value property" (the builder reads it). Mirrors Apple's
    // `ProgressView.initialValue`, which returns `model.value as? Double`.
    override fun initialValue(element: ActionUIElement): Any? = null

    @Composable
    override fun BuildView(element: ActionUIElement, modifier: Modifier) {
        val props = element.properties
        val logger = LocalActionUILogger.current
        val title = props?.stringProperty("title")
        val actionID = props?.stringProperty("actionID")
        // remember()ed on the properties: this composable recomposes on every
        // progress tick (that is what the value bridge and states["progress"]
        // are for), and resolving on each pass would log the invalid-style
        // warning hundreds of times for one bad string.
        val style = remember(props) {
            resolveProgressViewStyle(props?.stringProperty("progressViewStyle"), logger)
        }

        // Effective progress (Apple precedence): states["progress"] override >
        // the value bridge (a host write) > the static `value` property. Null at
        // every level -> indeterminate.
        val viewModel = LocalWindowModel.current?.viewModels?.get(element.id)
        val stateProgress = viewModel?.states?.get(PROGRESS_STATE_KEY) as? Double
        val runtimeValue = viewModel?.value as? Double
        val value = stateProgress ?: runtimeValue ?: props?.numberProperty("value")
        val total = props?.numberProperty("total") ?: if (value != null) 1.0 else null
        val determinate = value != null && value >= 0.0 &&
            total != null && total > 0.0 && value <= total
        val fraction = if (determinate) (value!! / total!!).toFloat() else 0f

        val rootModifier = if (actionID != null) {
            modifier.clickable(enabled = LocalActionUIEnabled.current) {
                ActionUIModel.actionHandler(actionID, viewID = element.id, viewPartID = 0)
            }
        } else {
            modifier
        }

        // SwiftUI `.tint` colors the progress indicator; an inherited tint maps
        // to the bar's / spinner's color.
        val tint = LocalActionUITint.current

        // Which SHAPE to draw is now a separate question from whether the
        // progress is known. "automatic" keeps the historical pairing.
        val linear = when (style) {
            ProgressViewStyle.Linear -> true
            ProgressViewStyle.Circular -> false
            ProgressViewStyle.Automatic -> determinate
        }

        Column(modifier = rootModifier) {
            if (title != null) M3Text(text = title)
            when {
                linear && determinate && tint != null ->
                    LinearProgressIndicator(progress = { fraction }, color = tint)
                linear && determinate -> LinearProgressIndicator(progress = { fraction })
                linear && tint != null -> LinearProgressIndicator(color = tint)
                linear -> LinearProgressIndicator()
                determinate && tint != null ->
                    CircularProgressIndicator(progress = { fraction }, color = tint)
                determinate -> CircularProgressIndicator(progress = { fraction })
                tint != null -> CircularProgressIndicator(color = tint)
                else -> CircularProgressIndicator()
            }
        }
    }
}

/** The ProgressView shapes the Android renderer supports. */
internal enum class ProgressViewStyle { Automatic, Linear, Circular }

/**
 * Maps the JSON `progressViewStyle` to a [ProgressViewStyle]. `null`/`"automatic"`
 * keeps the determinate-bar / indeterminate-spinner pairing; `"linear"` and
 * `"circular"` force the shape. Any other value warns and falls back to
 * automatic, parity with the Apple `validateProperties` default.
 */
internal fun resolveProgressViewStyle(raw: String?, logger: ActionUILogger): ProgressViewStyle = when (raw) {
    null, "automatic" -> ProgressViewStyle.Automatic
    "linear" -> ProgressViewStyle.Linear
    "circular" -> ProgressViewStyle.Circular
    else -> {
        logger.log(
            "ProgressView progressViewStyle '$raw' is not recognized; using automatic",
            LoggerLevel.warning
        )
        ProgressViewStyle.Automatic
    }
}

File: ActionUIWeb/src/Views/ProgressView.js
// ProgressView.js — ProgressView element.
// Web analog of ActionUI/Views/ProgressView.swift (and ActionUIAndroid Views/ProgressView.kt).
//
// A progress indicator, rendered as a native <progress>. It is **determinate**
// (a filled bar) when a valid value is supplied and value <= total, otherwise
// **indeterminate** (the browser's animated bar — the web counterpart of the
// SwiftUI/Material spinner). valueType is "double": a host reads/writes the
// progress with get/setDouble(id); setting the value to null reverts to
// indeterminate, mirroring Apple's states["progress"] = nil.
//
// Properties (mirroring ProgressView.swift):
//   value    current progress, 0.0…total (Double). Omit (or supply an invalid /
//            out-of-range value) for an indeterminate indicator.
//   total    maximum (positive Double); defaults to 1.0 when value is present.
//   title    optional label shown above the indicator.
//   progressViewStyle  "automatic" | "linear" | "circular". The web has only one
//            native indicator, <progress>, and it is LINEAR in both the
//            determinate and the indeterminate state - so this port already
//            renders what "linear" asks for and the property is accepted here
//            for portability. "circular" has no native analog and is not ported:
//            it warns and stays linear, the same warn-and-fall-back the Picker
//            and Toggle ports use for a style a platform cannot honor.
//   actionID dispatched on tap/click (like Button), with no context.

import { register } from "../Common/ActionUIRegistry.js";
import { markHandlesAction } from "../Common/ModifierResolver.js";

register("ProgressView", {
    valueType: "double",

    // Mirrors ProgressView.swift validateProperties (warning text included verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };

        // value: a non-negative number, else dropped (⇒ indeterminate).
        if (validated.value !== undefined && !(typeof validated.value === "number" && validated.value >= 0.0)) {
            logger.log("ProgressView value must be a non-negative Double; defaulting to nil", "warning");
            delete validated.value;
        }

        // total: a positive number, else dropped.
        if (validated.total !== undefined && !(typeof validated.total === "number" && validated.total > 0.0)) {
            logger.log("ProgressView total must be a positive Double; defaulting to nil", "warning");
            delete validated.total;
        }

        // title: must be a String, else dropped.
        if (validated.title !== undefined && typeof validated.title !== "string") {
            logger.log("ProgressView title must be a String; defaulting to nil", "warning");
            delete validated.title;
        }

        // progressViewStyle: one of the three names, else dropped.
        if (validated.progressViewStyle !== undefined) {
            const validStyles = ["automatic", "linear", "circular"];
            if (typeof validated.progressViewStyle !== "string") {
                logger.log("ProgressView progressViewStyle must be a String; defaulting to nil", "warning");
                delete validated.progressViewStyle;
            } else if (!validStyles.includes(validated.progressViewStyle)) {
                logger.log(`ProgressView progressViewStyle '${validated.progressViewStyle}' must be one of ${validStyles}; defaulting to nil`, "warning");
                delete validated.progressViewStyle;
            }
        }

        return validated;
    },

    // Seeds the determinate value; absent ⇒ undefined ⇒ no seed (indeterminate).
    initialValue: (element, properties) => properties.value,

    buildView: (element, properties, ctx) => {
        // total defaults to 1.0 when a value is present — the documented behavior,
        // matching the Android port. (Apple's buildView has a latent quirk, noted
        // by its own TODO, where an absent total falls through to indeterminate;
        // we follow the documented/Android semantics. See Web_Porting_Notes.md.)
        const total = properties.total ?? 1.0;
        let current = (typeof properties.value === "number") ? properties.value : null;

        const isDeterminate = (value) => value !== null && Number.isFinite(value) && value >= 0 && value <= total;

        // <progress> is linear in both states, so "linear" and "automatic" are
        // already what this builds. "circular" has no native analog on the web;
        // warn and stay linear rather than silently rendering something the JSON
        // did not ask for. The style is recorded on the element either way, as
        // Picker does with data-aui-picker-style, so a stylesheet can hook it.
        const style = properties.progressViewStyle;
        if (style === "circular") {
            ctx.logger.log("ProgressView progressViewStyle 'circular' is not ported on the web; using the linear indicator", "warning");
        }

        const bar = document.createElement("progress");
        bar.className = "aui-progress-bar";
        if (typeof style === "string") {
            bar.dataset.auiProgressViewStyle = style;
        }
        bar.max = total;
        // A <progress> with no value attribute renders as indeterminate; setting
        // .value makes it determinate.
        const renderValue = () => {
            if (isDeterminate(current)) {
                bar.value = current;
            } else {
                current = null;
                bar.removeAttribute("value");
            }
        };
        renderValue();

        // A title wraps the bar in a column, matching ProgressView(title, …) and
        // the Android Column; otherwise the bare bar is the root.
        let root = bar;
        if (properties.title) {
            root = document.createElement("div");
            root.className = "aui-progress";
            const label = document.createElement("span");
            label.className = "aui-progress-title";
            label.textContent = properties.title;
            root.append(label, bar);
        }

        // actionID fires on tap/click, like Button (no context), on the root.
        if (typeof properties.actionID === "string") {
            markHandlesAction(root);
            root.addEventListener("click", () => {
                ctx.model.dispatchAction(properties.actionID, element.id);
            });
        }

        if (element.id > 0) {
            ctx.model.bind(element.id, {
                // null when indeterminate (the host can branch on it); a number
                // otherwise.
                getValue: () => (isDeterminate(current) ? current : null),
                setValue: (value) => {
                    // A valid number ⇒ determinate; null/invalid ⇒ indeterminate
                    // (the web analog of setting states["progress"] = nil).
                    const n = Number(value);
                    current = (value !== null && value !== undefined && Number.isFinite(n)) ? n : null;
                    renderValue();
                },
            });
        }
        return root;
    },
});
